### DISCUSSION TITLE
`MissingGreenlet` when accessing relationship collection of new instance after flush()

### DISCUSSION THREAD

**Discussion by dexter-dopping-ekco** [sqlalchemy/sqlalchemy#8913] [Usage Questions] [answered]
Hello! My problem occurs in the following snippet:

```python
async with Session() as session:
    new_customer = Customer(name="John")

    session.add(new_customer)
    await session.flush()

    # Here it throws an sqlalchemy.exc.MissingGreenlet exception
    print(new_customer.id, new_customer.giraffe_accesses)
```

`new_customer.giraffe_accesses` is a relationship to the `GiraffeAccess` model.

This is strange to me because it doesn't happen when I make the following change:

```patch
 async with Session() as session:
     new_customer = Customer(name="John")
+    new_customer.giraffe_accesses.append(GiraffeAccess(giraffe_id=1))
 
     session.add(new_customer)
     await session.flush()
 
     # Now it doesn't raise an exception
     print(new_customer.id, new_customer.giraffe_accesses)
```

I'm using `expire_on_commit=False`. I don't understand why it needs to re-query the `new_customer.giraffe_accesses` items, especially the fact that it *only* happens when nothing is added to the list. There's a good chance I'm misunderstanding how it's supposed to work..

How can I make it so the code runs successfully no matter if items were added to `new_customer.giraffe_accesses` or not? Preferable, I wouldn't fetch the data again from the database because it doesn't seem necessary but maybe I'm missing something.

Here the full sample:

```python
import asyncio

from sqlalchemy import (
    Column,
    Integer,
    String,
    ForeignKey,
)
from sqlalchemy.orm import declarative_base, relationship, sessionmaker
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession


Base = declarative_base()
metadata = Base.metadata

class GiraffeAccess(Base):
    __tablename__ = "giraffe_accesses"
    giraffe_id = Column(Integer, primary_key=True)
    customer_id = Column(Integer, ForeignKey("customers.id"), primary_key=True)

class Customer(Base):
    __tablename__ = "customers"
    id = Column(Integer, primary_key=True)
    name = Column(String(100))
    giraffe_accesses = relationship("GiraffeAccess")


engine = create_async_engine(
    "postgresql+asyncpg://giraffe_user:grffs@localhost:5432/giraffes"
)
Session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)

async def setup():
    async with engine.begin() as conn:
        await conn.run_sync(metadata.drop_all)
        await conn.run_sync(metadata.create_all)


async def main():

    # This works
    async with Session() as session:
        new_customer = Customer(name="John")
        new_customer.giraffe_accesses.append(GiraffeAccess(giraffe_id=1))

        session.add(new_customer)
        await session.flush()

        print(new_customer.id, new_customer.giraffe_accesses)

    # This doesn't work
    async with Session() as session:
        new_customer = Customer(name="John")

        session.add(new_customer)
        await session.flush()

        # Here it throws an sqlalchemy.exc.MissingGreenlet exception
        print(new_customer.id, new_customer.giraffe_accesses)


async def run():
    await setup()
    await main()

asyncio.run(run())
```

Using:

- SQLAlchemy==1.4.44
- asyncpg==0.27.0
- Python 3.10.8
- WSL Debian 11, running the code and hosting the database
- PostgreSQL 14.5

**Comment by zzzeek** [comment, marked as answer]
> Hello! My problem occurs in the following snippet:
> 
>     new_customer = Customer(name="John")

When you run the above line, that's fine, but "new_customer.giraffe_accesses" is entirely empty, and there is no list present there.


>     session.add(new_customer)
>     await session.flush()

also fine, however, there is nothing loaded whatsoever into "new_customer.giraffe_accesses", and there is no list present there.  the flush process does not eagerly load unloaded relationships on target objects, this would be wasteful (imagine if your class had 20 collections on it and you flushed 100 objects.   why would flush take the time to load all those collections when it doesnt need to?)

> 
>     # Here it throws an sqlalchemy.exc.MissingGreenlet exception
>     print(new_customer.id, new_customer.giraffe_accesses)

this is because asyncio doesn't support lazy loading.  Background on this is here: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession

> This is strange to me because it doesn't happen when I make the following change:
> 
> ```diff
>  async with Session() as session:
>      new_customer = Customer(name="John")
> +    new_customer.giraffe_accesses.append(GiraffeAccess(giraffe_id=1))

So by accessing `new_customer.giraffe_accesses`, because the object is transient, a new, empty list is created for you to append objects towards.  Then you append your object into it.  `new_customer` now has `new_customer.giraffe_accesses` present with the object you placed into it.


> 
>      session.add(new_customer)
>      await session.flush()

the flush process does as it did before, flushing "new_customer", and it also sees that you added a pending object into `new_customer.giraffe_accesses`, so it flushes that too.   It does not do anything differently from before, however.   `new_customer.giraffe_accesses` is now present because you put objects into it.   

>  
>      # Now it doesn't raise an exception
>      print(new_customer.id, new_customer.giraffe_accesses)

lazy loading isn't needed here because you populated the `new_customer.giraffe_accesses` collection yourself.

> ```
> I'm using `expire_on_commit=False`. 

your example accesses the collection before the commit occurs in any case, so this is not relevant.

> I don't understand why it needs to re-query the `new_customer.giraffe_accesses` items, 

it never queried it in the first place, so it is not "re" querying anything.

> especially the fact that it _only_ happens when nothing is added to the list. 

it happens when your code did not access this attribute in any way, so SQLAlchemy leaves it unaffected (keeping in mind, objects with 20 collections each and thousands of objects, SQLAlchemy does not eagerly access all these collections if it isn't told to, as this would be very wasteful).


> 
> How can I make it so the code runs successfully no matter if items were added to `new_customer.giraffe_accesses` or not? 

include giraffe_accesses in your constructor, is the most succinct and explicit way:

```py
new_customer = Customer(name="John", giraffe_accesses=[])
```

that way you are not relying on implicit behaviors.

**Comment by dexter-dopping-ekco** [comment]
Hey Mike,

It seems very similar to using `server_default` for a column, except in this scenario I expected SQLAlchemy to "know" that `giraffe_accesses` would have to be an empty list.

Initialising `giraffe_accesses` to `[]` accomplishes exactly what I was looking for. I'm having a hard time coming up with a scenario where I wouldn't want this.

 Thank you so much for the detailed explanation.

Kind regards,
Dexter
